Foundation/NSDecimal.swift

```diff
@@ -56,74 +187,286 @@ extension Decimal {
+    }
+    public init(_ value: Int64) {
+        self.init(Double(value))
+    }
+    public init(_ value: UInt) {
+        self.init(UInt64(value))
+    }
+    public init(_ value: Int) {
+        self.init(Int64(value))
+    }
+    public var isSignalingNaN: Bool {
+        return false
+    }
+    public static var nan: Decimal {
+        return quietNaN
+    }
+    public static var quietNaN: Decimal {
+        var quiet = Decimal()
+        quiet._isNegative = 1
+        return quiet
+    }
+    public var floatingPointClass: FloatingPointClassification {
+        if _length == 0 && _isNegative == 1 {
+            return .quietNaN
+        } else if _length == 0 {
+            return .positiveZero
+        }
+        if _isNegative == 1 {
+            return .negativeNormal
+        } else {
+            return .positiveNormal
+        }
+    }
+    public var isSignMinus: Bool {
+        return _isNegative != 0
+    }
+    public var isNormal: Bool {
+        return !isZero && !isInfinite && !isNaN
+    }
+    public var isFinite: Bool {
+        return !isNaN
+    }
+    public var isZero: Bool {
+        return _length == 0 && _isNegative == 0
+    }
+    public var isSubnormal: Bool {
+        return false
+    }
+    public var isInfinite: Bool {
+        return false
+    }
+    public var isNaN: Bool {
+        return _length == 0 && _isNegative == 1
+    }
+    public var isSignaling: Bool {
+        return false
+    }
 }
 
 extension Decimal : CustomStringConvertible {
     public init?(string: String, locale: Locale? = nil) {
-        // Avoid a compiler warning for now by not calling NSUnimplemented
-        return nil
+        let scan = Scanner(string: string)
+        var theDecimal = Decimal()
+        if !scan.scanDecimal(&theDecimal) {
+            return nil
+        }
+        self = theDecimal
+    }
+    public var description: String {
+        var val = self
+        return NSDecimalString(&val, nil)
     }
-    public var description: String { NSUnimplemented() }
 }
 
-public func NSDecimalIsNotANumber(_ dcm: UnsafePointer<Decimal>) -> Bool { NSUnimplemented() }
-
+public func NSDecimalIsNotANumber(_ dcm: UnsafePointer<Decimal>) -> Bool {
+	return dcm.pointee.isNaN
+}
 
 /***************	Operations		***********/
-public func NSDecimalCopy(_ destination: UnsafeMutablePointer<Decimal>, _ source: UnsafePointer<Decimal>) { NSUnimplemented() }
+public func NSDecimalCopy(_ destination: UnsafeMutablePointer<Decimal>, _ source: UnsafePointer<Decimal>) {
+    destination.pointee.__lengthAndFlags = source.pointee.__lengthAndFlags
+    destination.pointee.__exponent = source.pointee.__exponent
+    destination.pointee.__reserved = source.pointee.__reserved
+    destination.pointee._mantissa = source.pointee._mantissa
+}
 
-public func NSDecimalCompact(_ number: UnsafeMutablePointer<Decimal>) { NSUnimplemented() }
+public func NSDecimalCompact(_ number: UnsafeMutablePointer<Decimal>) {
+    number.pointee.compact()
```

Comment:
Using the UnsafeMutablePointer is actually pretty painful. You can't (for example) have 'var d = number.pointee; d.doStuff()' because it ends up with a different copy. So everywhere you have the functions you need to do number.pointee to use the passed in version. In addition, it's not being safe for the sake of it, but rather adopting an older API.

Finally you can't tell what is being mutated and what isn't with these APIs; you're just passing an opaque pointer. This is tedious when doing non-mutating code (e.g. the NSDecimalIsNotANumber) because you have can't pass in a 'let' constant:

  1> import Foundation
  2> let one = Decimal(1)
one: Decimal = {
  _mantissa = {
    0 = 8448
    1 = 0
    2 = 1
    3 = 0
    4 = 0
    5 = 0
    6 = 0
    7 = 0
  }
}
  3> NSDecimalIsNotANumber(one)
error: repl.swift:3:23: error: cannot convert value of type 'Decimal' to expected argument type 'UnsafePointer<Decimal>'
NSDecimalIsNotANumber(one)
                      ^~~


  3> NSDecimalIsNotANumber(&one)
error: repl.swift:3:23: error: cannot pass immutable value as inout argument: 'one' is a 'let' constant
NSDecimalIsNotANumber(&one)
                      ^~~~

Having those as the API is probably a bad thing to do; in fact, I'd argue that the NSDecimalXxxx functions should be marked as deprecated. Right now the simple (non-mutating) ones can be more efficiently implemented (read: not taking an unsafe pointer) by doing it the reverse way from the way that Objective-C did it, and having the functions as backwards compatible for those that need it.